src/main/scala/chisel3/util/SRAM.scala

```diff
@@ -470,54 +472,140 @@ object SRAM {
+
+    // user-facing interface into the SRAM
+    val sramIntfType = new SRAMInterface(size, tpe, numReadPorts, numWritePorts, numReadwritePorts, isVecMem)
+    val _out = Wire(sramIntfType)
     _out._underlying = Some(HasTarget(mem))
 
-    for ((clock, port) <- readPortClocks.zip(_out.readPorts)) {
-      port.data := mem.read(port.address, port.enable, clock)
+    // create actual ports into firrtl memory
+    val firrtlReadPorts:  Seq[FirrtlMemoryReader[_]] = sramIntfType.readPorts.map(new FirrtlMemoryReader(_))
+    val firrtlWritePorts: Seq[FirrtlMemoryWriter[_]] = sramIntfType.writePorts.map(new FirrtlMemoryWriter(_))
+    val firrtlReadwritePorts: Seq[FirrtlMemoryReadwriter[_]] =
+      sramIntfType.readwritePorts.map(new FirrtlMemoryReadwriter(_))
+
+    // set references to firrtl memory ports
+    val firrtlReadPortNames: Seq[String] = firrtlReadPorts.zipWithIndex.map {
+      case (p, idx) =>
+        val name = s"R$idx"
+        p.setRef(Slot(Node(mem), name))
+        name
+    }
+    val firrtlWritePortNames: Seq[String] = firrtlWritePorts.zipWithIndex.map {
+      case (p, idx) =>
+        val name = s"W$idx"
+        p.setRef(Slot(Node(mem), name))
+        name
+    }
+    val firrtlReadwritePortNames: Seq[String] = firrtlReadwritePorts.zipWithIndex.map {
+      case (p, idx) =>
+        val name = s"RW$idx"
+        p.setRef(Slot(Node(mem), name))
+        name
+    }
+
+    // set bindings of firrtl memory ports
+    firrtlReadPorts.foreach(_.bind(SramPortBinding(Builder.forcedUserModule, Builder.currentWhen)))
+    firrtlWritePorts.foreach(_.bind(SramPortBinding(Builder.forcedUserModule, Builder.currentWhen)))
+    firrtlReadwritePorts.foreach(_.bind(SramPortBinding(Builder.forcedUserModule, Builder.currentWhen)))
+
+    // bind type so that memory type can get converted to FIRRTL
+    val boundType = tpe.cloneTypeFull
+    boundType.bind(FirrtlMemTypeBinding(mem))
+
+    // create FIRRTL memory
+    Builder.pushCommand(
+      FirrtlMemory(
+        sourceInfo,
+        mem,
+        boundType,
+        size,
+        firrtlReadPortNames,
+        firrtlWritePortNames,
+        firrtlReadwritePortNames
+      )
+    )
+
+    // connect firrtl memory ports to user-facing interface
+    _out.readPorts.zip(firrtlReadPorts).foreach {
+      case (memReadPort, firrtlReadPort) =>
+        firrtlReadPort.addr := memReadPort.address
+        firrtlReadPort.clk := Module.clock
+        memReadPort.data := firrtlReadPort.data.asInstanceOf[Data]
+        firrtlReadPort.en := memReadPort.enable
     }
+    _out.writePorts.zip(firrtlWritePorts).foreach {
+      case (memWritePort, firrtlWritePort) =>
+        firrtlWritePort.addr := memWritePort.address
+        firrtlWritePort.clk := Module.clock
+        firrtlWritePort.data.asInstanceOf[Data] := memWritePort.data
+        firrtlWritePort.en := memWritePort.enable
+        assignMask(memWritePort.data, memWritePort.mask, firrtlWritePort.getRef, "mask")
+    }
```

Comment:
```suggestion
    for((memWritePort, firrtlWritePort) <- _out.writePorts.zip(firrtlWritePorts)) {
      firrtlWritePort.addr := memWritePort.address
      firrtlWritePort.clk := Module.clock
      firrtlWritePort.data.asInstanceOf[Data] := memWritePort.data
      firrtlWritePort.en := memWritePort.enable
      assignMask(memWritePort.data, memWritePort.mask, firrtlWritePort.getRef, "mask")
    }
```
This is a not a strongly held position, but for simple foreaches like this, I prefer `for` style since it reduces a level of indention

Reply:
oh cool, that does look nice! I have never seen or used this pattern before :x